### PR TITLE
Refactor LoadMD data indexing and memory algorithms to speed up processing

### PR DESCRIPTION
**Pull request by williamfgc**
**Description of work.**

Changes targeting LoadMD algorithms efficiency.
Refactor class hierarchy to inherit from `Mantid::Kernel::NexusHDF5Descriptor` and resuse metadata indexing.
Refactor metadata index processing using NexusHDF5Descriptor to avoid regenerating std::map at every level.
Changes in memory allocation reusing containers as much as possible.
Reported speed ups on 20% on average on ORNL SNS/HFIR production system for 180 experiments, it also reduces system variability effects.

**To test:**
Refactoring should pass existing CI tests, including system tests.

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

File: Framework/API/inc/MantidAPI/ExperimentInfo.h [hunk 1]
```diff
@@ -120,6 +121,11 @@ class MANTID_API_DLL ExperimentInfo {
   void saveExperimentInfoNexus(::NeXus::File *file, bool saveLegacyInstrument = true) const;
   /// Saves this experiment description to the open NeXus file
   void saveExperimentInfoNexus(::NeXus::File *file, bool saveInstrument, bool saveSample, bool saveLogs) const;
+
+  void loadExperimentInfoNexus(const std::string &nxFilename, ::NeXus::File *file, std::string &parameterStr,
+                               const std::shared_ptr<Mantid::Kernel::NexusHDF5Descriptor> &fileInfo,
```
Comment (quantumsteve) [inline Framework/API/inc/MantidAPI/ExperimentInfo.h:126]: Can we pass fileInfo as a reference (`Mantid::Kernel::NexusHDF5Descriptor &`) in this and similar functions?

File: Framework/DataObjects/src/MDBoxFlatTree.cpp [hunk 2]
```diff
@@ -378,6 +379,90 @@ void MDBoxFlatTree::saveExperimentInfos(::NeXus::File *const file, const API::IM
   }
 }
 
+void MDBoxFlatTree::loadExperimentInfos(::NeXus::File *const file, const std::string &filename,
```
Comment (quantumsteve) [inline Framework/DataObjects/src/MDBoxFlatTree.cpp:382]: Can we factor our portions of this member function that are the same as in `void MDBoxFlatTree::loadExperimentInfos(::NeXus::File *const file, const std::string &filename,
                                        const std::shared_ptr<Mantid::API::MultipleExperimentInfos> &mei, bool lazy)`?

File: Framework/Kernel/src/PropertyNexus.cpp [hunk 3]
```diff
@@ -108,6 +108,107 @@ std::unique_ptr<Property> makeStringProperty(::NeXus::File *file, const std::str
...
+/** Opens a NXlog group in a nexus file and
+ * creates the correct Property object from it. Overlead that
+ *
+ * @param file :: NXS file handle
+ * @param group :: name of NXlog group to open
+ * @param fileDescriptor :: HDF5 Nexus file descriptor (metadata)
+ * @param prefix :: current absolute target prefix
+ * @return Property pointer
+ */
+std::unique_ptr<Property> loadProperty(::NeXus::File *file, const std::string &group,
```
Comment (quantumsteve) [inline Framework/Kernel/src/PropertyNexus.cpp:121]: Can we factor our portions of this member function that are the same as in `std::unique_ptr<Property> loadProperty(::NeXus::File *file, const std::string &group)`?

File: Framework/API/inc/MantidAPI/ExperimentInfo.h (hunk 1 above)
Comment (williamfgc) [reply Framework/API/inc/MantidAPI/ExperimentInfo.h:126]: Good call.

**Comment by peterfpeterson** [comment]
For the comment "reported speed ups on 20% on average on ORNL SNS/HFIR production system" can you give more details? e.g. before and after timing, and characteristics of files.

**Comment by williamfgc** [comment]
@quantumsteve thanks for the review, I addressed your points. 
@peterfpeterson please see the attached image running on analysis.sns.gov :

Screen Shot 2021-09-14 at 7 46 37 PM

**Comment by williamfgc** [comment]
It should be good to go.

**Review by peterfpeterson — APPROVED** [review]

**Pull request merged**